### PR TITLE
CAAM driver minor fixes

### PR DESCRIPTION
**Pull request by clementfaure**
Hello,

Here are some patches for the CAAM driver. Most of these patches fix minor issues in the CAAM.
However, the patch fixing the issue I found last week in xtest is the following : `drivers: caam: fix missing output additional length`

Thanks,
Clement

### CODE REVIEW

**Review by jforissier — COMMENTED** [review]
One general comment: when a commit is a fix for a previous commit that is already part of a released version of OP-TEE, it is nice to add a `Fixes:` tag citing the commit that is being fixed. The reason is to make it easier for downstream projects to select and pick patches based on what they have in their tree.

Consider for example commit 1d88c0c03f3bed5776739e459a598a011baa5207 in release 3.14.0 which fixes 59ac3801b756 introduced in 3.13.0.

Now, sometimes we add `Fixes:` tags for bugs introduced in the same release (probably less useful) or sometimes we forget to add them... but I think we should try to do it more consistently. `git blame <file>`, `git show <commit>`, `git tag --contains <commit>` are helpful commands.

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 1]
```diff
@@ -73,9 +73,9 @@ uint8_t caam_hal_ctrl_hash_limit(vaddr_t baseaddr)
 
 bool caam_hal_ctrl_splitkey_support(vaddr_t baseaddr)
 {
-	uint32_t val = io_caam_read32(baseaddr + CAAMVID_MS);
+	uint32_t val = io_caam_read32(baseaddr + CTPR_LS);
 
-	return GET_CAAMVID_MS_MAJ_REV(val) >= 3;
+	return !!GET_CTPR_LS_SPLIT_KEY(val);
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:78]: I will remove it

File: core/drivers/crypto/crypto_api/acipher/ecc.c [hunk 2]
```diff
@@ -98,16 +98,19 @@ static TEE_Result ecc_generate_keypair(struct ecc_keypair *key,
...
 	struct drvcrypt_ecc *ecc = NULL;
+	size_t key_size = 0;
 
 	/* Check input parameters */
 	if (!key) {
 		CRYPTO_TRACE("Parameters error key is NULL");
 		return TEE_ERROR_BAD_PARAMETERS;
 	}
 
+	key_size = get_ecc_key_size_bytes(key->curve) * 8;
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/crypto_api/acipher/ecc.c:109]: Sorry about that, that's the result of squashing two commits .. I will split this commits into two.

**Comment by clementfaure** [comment]
Thanks for your feedback @jforissier 
It's true that we tend to forget about adding `Fixes: ` to our commit description. We will try on our side to stick to this good practice.
I added `Fixes: ` to these commits. Most of them fix initial CAAM driver patches. I also addressed your two comments. Let me know if that works for you.

File: core/drivers/crypto/crypto_api/acipher/ecc.c [hunk 3]
```diff
@@ -98,16 +98,19 @@ static TEE_Result ecc_generate_keypair(struct ecc_keypair *key,
...
+
 	ecc = drvcrypt_get_ops(CRYPTO_ECC);
 	if (ecc)
-		ret = ecc->gen_keypair(key, get_ecc_key_size_bytes(key->curve));
+		ret = ecc->gen_keypair(key, key_size);
 
 	CRYPTO_TRACE("ECC Keypair (%zu bits) generate ret = 0x%" PRIx32,
 		     size_bits, ret);
```
Comment (jforissier) [inline core/drivers/crypto/crypto_api/acipher/ecc.c:116]: s/size_bits/key_size/?

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 4]
```diff
@@ -75,7 +75,7 @@ bool caam_hal_ctrl_splitkey_support(vaddr_t baseaddr)
 {
 	uint32_t val = io_caam_read32(baseaddr + CTPR_LS);
 
-	return !!GET_CTPR_LS_SPLIT_KEY(val);
+	return !GET_CTPR_LS_SPLIT_KEY(val);
```
Comment (jforissier) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:78]: Oops! You left one `!`.

**Review by jforissier — COMMENTED** [review]
With the two things below addressed, LGTM.
`Acked-by: Jerome Forissier <jerome@forissier.org>`

File: core/drivers/crypto/crypto_api/acipher/ecc.c (hunk 3 above)
Comment (clementfaure) [reply core/drivers/crypto/crypto_api/acipher/ecc.c:116]: Oops ... fixed

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 4 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:78]: Oops ... fixed

**Comment by clementfaure** [comment]
Force-push:
- Squash review commits
- Address the last two comments
- Applied tags

File: core/drivers/crypto/crypto_api/acipher/ecc.c [hunk 5]
```diff
@@ -98,19 +98,22 @@ static TEE_Result ecc_generate_keypair(struct ecc_keypair *key,
 {
 	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
 	struct drvcrypt_ecc *ecc = NULL;
+	size_t key_size = 0;
```
Comment (etienne-lms) [inline core/drivers/crypto/crypto_api/acipher/ecc.c:101]: for consistency, prefer rename to `key_size_bits`.

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c [hunk 6]
```diff
@@ -75,7 +75,7 @@ bool caam_hal_ctrl_splitkey_support(vaddr_t baseaddr)
 {
 	uint32_t val = io_caam_read32(baseaddr + CTPR_LS);
 
-	return !GET_CTPR_LS_SPLIT_KEY(val);
+	return GET_CTPR_LS_SPLIT_KEY(val);
```
Comment (etienne-lms) [inline core/drivers/crypto/caam/hal/common/hal_ctrl.c:78]: should belong to another commit.

**Review by etienne-lms — COMMENTED** [review]
comments for commit "drivers: crypto: fix ECC key size bytes to bits conversion".

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 4 above)
Comment (etienne-lms) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c]: note it's been fixed in the wrong commit.

File: core/drivers/crypto/crypto_api/acipher/ecc.c (hunk 5 above)
Comment (clementfaure) [reply core/drivers/crypto/crypto_api/acipher/ecc.c:101]: I agree

File: core/drivers/crypto/caam/hal/common/hal_ctrl.c (hunk 6 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/hal/common/hal_ctrl.c:78]: My bad, I will fix it

File: core/drivers/crypto/caam/utils/utils_dmaobj.c [hunk 7]
```diff
@@ -915,8 +921,14 @@ size_t caam_dmaobj_copy_ltrim_to_orig(struct caamdmaobj *obj)
...
 
 do_copy:
-	if (off < obj->orig.length)
-		dst_rlen = obj->orig.length - off;
+	/*
+	 * The maximum data size to copy cannot exceed the output buffer size
+	 * (obj->orig.length) and cannot exceed the data processed by the
+	 * CAAM (obj->sgtbuf.length).
+	 */
+	dst_rlen = MIN(obj->orig.length, obj->sgtbuf.length);
```
Comment (etienne-lms) [inline core/drivers/crypto/caam/utils/utils_dmaobj.c:929]: this changes `dst_rlen` value. I think test at line 938 is corrupted by this change.

**Review by etienne-lms — COMMENTED** [review]
comment for commit "drivers: caam: dma: fix copy data size"

**Comment by clementfaure** [comment]
Force-push:
* Split commit `drivers: crypto: fix ECC key size bytes to bits conversion` to fix the CAAM split key in the appropriate commit

File: core/drivers/crypto/caam/utils/utils_dmaobj.c (hunk 7 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_dmaobj.c:929]: What do you mean by corrupted? To me, this test still looks valid

Comment (etienne-lms) [reply core/drivers/crypto/caam/utils/utils_dmaobj.c:929]: With previous implementation `dst_rlen` was 0 when  `off >= obj->orig.length`. Here it likely always gets non 0 value. It could cause trouble IMO.

**Comment by ruchi393** [comment]
`Acked-by: Ruchika Gupta <ruchika.gupta@linaro.org>`

File: core/drivers/crypto/caam/utils/utils_dmaobj.c (hunk 7 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_dmaobj.c:929]: You are right, the change might have an impact on the following `if (!dst_rlen) {` statement.
In the end, this change in the function `caam_dmaobj_copy_ltrim_to_orig()` is not needed. This function is only used in the RSA crypto operation and the `obj->orig.length` cannot be smaller than `obj->sgtbuf.length` because of the SGT allocation for the RSA. Thus, the modification done is not needed.

**Comment by clementfaure** [comment]
Force-push:
- Applied @ruchi393 tags
- The CI failed for some reasons so I rebased the PR.

@etienne-lms Do you have any other comments/questions?

File: core/drivers/crypto/caam/utils/utils_dmaobj.c (hunk 7 above)
Comment (clementfaure) [reply core/drivers/crypto/caam/utils/utils_dmaobj.c]: @etienne-lms Gentle reminder!

File: core/drivers/crypto/caam/hal/common/registers/rng_regs.h [hunk 8]
```diff
@@ -22,27 +22,88 @@
...
 
 /* Seed Control */
-#define TRNG_SDCTL					0x0610
+#define TRNG_SDCTL			0x0610
 #define BM_TRNG_SDCTL_ENT_DLY		SHIFT_U32(0xFFFF, 16)
 #define GET_TRNG_SDCTL_ENT_DLY(val)	(((val) & BM_TRNG_SDCTL_ENT_DLY) >> 16)
 #define TRNG_SDCTL_ENT_DLY(val)		SHIFT_U32(((val) & 0xFFFF), 16)
+#define BM_TRNG_SDCTL_SAMP_SIZE		SHIFT_U32(0xFFFF, 0)
+#define GET_TRNG_SDCTL_SAMP_SIZE(val)	((val) & (BM_TRNG_SDCTL_SAMP_SIZE))
+#define TRNG_SDCTL_SAMP_SIZE(val)	SHIFT_U32(((val) & (0xFFFF)), 0)
```
[the remark this answers is not in the document]
Comment (clementfaure) [reply core/drivers/crypto/caam/hal/common/registers/rng_regs.h:31]: Thanks for noticing it! I will remove these values.
I will also reformat `TRNG_SDCTL_SAMP_SIZE`

**Comment by clementfaure** [comment]
Force-push:
* Squash review commits
* Tag applied (@etienne-lms) Thanks!

**Comment by clementfaure** [comment]
Force-push:
* Rebase
